satpy/readers/amsr2_l2_gaasp.py

```diff
@@ -0,0 +1,286 @@
+
+logger = logging.getLogger(__name__)
+
+
+class GAASPFileHandler(BaseFileHandler):
+    """Generic file handler for GAASP output files."""
+
+    y_dims = (
+        'Number_of_Scans',
+    )
+    x_dims = (
+        'Number_of_hi_rez_FOVs',
+        'Number_of_low_rez_FOVs',
+    )
+    time_dims = (
+        'Time_Dimension',
+    )
+    is_gridded = False
+    dim_resolutions = {
+        'Number_of_hi_rez_FOVs': 5000,
+        'Number_of_low_rez_FOVs': 10000,
+    }
+
+    def __init__(self, filename, filename_info, filetype_info):
+        """Open the NetCDF file with xarray and prepare the Dataset for reading."""
+        super().__init__(filename, filename_info, filetype_info)
+        chunks = {dim_name: CHUNK_SIZE for dim_name in
+                  self.y_dims + self.x_dims + self.time_dims}
+        self.nc = xr.open_dataset(self.filename,
+                                  decode_cf=True,
+                                  mask_and_scale=False,
+                                  chunks=chunks)
+
+        if len(self.time_dims) == 1:
+            self.nc = self.nc.rename({self.time_dims[0]: 'time'})
+
+    @property
+    def start_time(self):
+        """Get start time of observation."""
+        time_str = self.nc.attrs['time_coverage_start']
+        return datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+    @property
+    def end_time(self):
+        """Get end time of observation."""
+        time_str = self.nc.attrs['time_coverage_end']
+        return datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+    @property
+    def sensor_names(self):
+        """Sensors who have data in this file."""
+        return {self.nc.attrs['instrument_name'].lower()}
+
+    @property
+    def platform_name(self):
+        """Name of the platform whose data is stored in this file."""
+        return self.nc.attrs['platform_name']
+
+    def _get_var_name_without_suffix(self, var_name):
+        var_suffix = self.filetype_info.get('var_suffix', "")
+        if var_suffix:
+            var_name = var_name[:-len(var_suffix)]
+        return var_name
+
+    def _scale_data(self, data_arr, attrs):
+        # handle scaling
+        # take special care for integer/category fields
+        scale_factor = attrs.pop('scale_factor', 1.)
+        add_offset = attrs.pop('add_offset', 0.)
+        scaling_needed = not (scale_factor == 1 and add_offset == 0)
+        if scaling_needed:
+            data_arr = data_arr * scale_factor + add_offset
+        return data_arr, attrs
+
+    @staticmethod
+    def _nan_for_dtype(data_arr_dtype):
+        # don't force the conversion from 32-bit float to 64-bit float
+        # if we don't have to
+        if data_arr_dtype.type == np.float32:
+            return np.float32(np.nan)
+        elif np.issubdtype(data_arr_dtype, np.timedelta64):
+            return np.timedelta64('NaT')
+        elif np.issubdtype(data_arr_dtype, np.datetime64):
+            return np.datetime64('NaT')
+        return np.nan
```

Comment:
Don't we have a function like this already defined in another reader, so that we could factorize some code?

Reply:
Not exactly like this, but yes we probably could.

Reply:
So I did some searching in satpy and there is no use exactly like this. The closest are some if statements specifically checking for integer versus float and assuming that any float data will be 32-bit floats. My files specifically have some timedelta fields so I needed these extra checks. I'm sure other readers could need this in the future. However, seeing as `where` operations and other basic arithmetic operations seem to now respect dtype precision (32-bit floats stay 32-bit floats, see my other comment) I'm not sure this will be needed by those other readers that are currently only using 32-bit floats.

Lastly, moving this to a utility function should probably be done in another PR as it might end up being a decent refactor for each effected reader.

Reply:
Agreed, thanks for double checking!